apps/homescreen/js/apps.js

```diff
@@ -155,218 +155,222 @@
+                this.startupMetadata = results.order;
+                this.settings.small = results.small;
+                this.settings.save();
+                return Promise.resolve();
+              }, (e) => {
+                console.error('Error running first-run script', e);
+                return Promise.resolve();
+              });
+            },
+            (e) => {
+              console.error('Failed to load first-run script');
               return Promise.resolve();
             });
+          } :
+          () => {
+            return this.metadata.getAll(result => {
+              this.startupMetadata.push(result);
+              this.metadataLoaded++;
+
+              // Process results in batches as they come in
+              var processResult = data => {
+                if (this.pendingIcons[data.id]) {
+                  var args = this.pendingIcons[data.id];
+                  delete this.pendingIcons[data.id];
+                  this.addAppIcon.apply(this, args);
+                }
+              };
+
+              if (!this.visualLoadComplete) {
+                processResult(result);
+                this.refreshGridSize();
+              } else if ((this.metadataLoaded % this.iconsPerPage) === 0) {
+                this.icons.freeze();
+                for (var entry of this.startupMetadata) {
+                  processResult(entry);
+                }
+                this.refreshGridSize();
+                this.icons.thaw();
+              }
+            }).then(Promise.resolve(),
+              e => {
+                console.error('Failed to retrieve metadata entries', e);
+                return Promise.resolve();
+              });
           },
           (e) => {
-            console.error('Failed to load first-run script');
+            console.error('Failed to initialise metadata db', e);
             return Promise.resolve();
-          });
-        } :
-        () => {
-          return this.metadata.getAll(result => {
-            this.startupMetadata.push(result);
-            this.metadataLoaded++;
-
-            // Process results in batches as they come in
-            var processResult = data => {
-              if (this.pendingIcons[data.id]) {
-                var args = this.pendingIcons[data.id];
-                delete this.pendingIcons[data.id];
-                this.addAppIcon.apply(this, args);
-              }
-            };
-
-            if (!this.visualLoadComplete) {
-              processResult(result);
-              this.refreshGridSize();
-            } else if ((this.metadataLoaded % this.iconsPerPage) === 0) {
-              this.icons.freeze();
-              for (var entry of this.startupMetadata) {
-                processResult(entry);
-              }
-              this.refreshGridSize();
-              this.icons.thaw();
+          }),
+
+        // Populate apps
+        new Promise((resolve, reject) => {
+          var request = navigator.mozApps.mgmt.getAll();
```

Comment:
Given it isn't related to the localstorage usage, I'd rather lump it in with the patches that fix up the places db stuff.